### PR TITLE
langserver: Support InitializationOptions

### PR DESCRIPTION
**Pull request by keegancsmith**
Fixes https://github.com/sourcegraph/go-langserver/issues/280

cc @lloiser @Contextualist @ramya-rao-a

### CODE REVIEW

File: langserver/handler.go [hunk 1]
```diff
@@ -73,7 +73,14 @@ type LangHandler struct {
...
 
-	Config Config // language handler configuration; must not change after handling has begun
+	// DefaultConfig is the default values used for configuration. It is
+	// combined with InitializationOptions after initialize. This should be
+	// set by LangHandler creators. Please read config instead.
+	DefaultConfig Config
+
+	// config is the language handler configuration. It is Default with
+	// InitializationOptions applied.
+	config *Config // pointer so we panic if someone reads before we set it.
```
Comment (lloiser) [inline langserver/handler.go]: config is the language handler configuration. It is a combination of `DefaultConfig` and `InitializationOptions`.

**Comment by Contextualist** [comment]
Is it better if we just take advantage of the standard `ClientCapabilities`?
i.e. 
`config.GocodeCompletionEnabled` -> `init.Capabilities.TextDocument.Completion`
`config.FuncSnippetEnabled` -> `init.Capabilities.TextDocument.Completion.SnippetSupport`

**Comment by keegancsmith** [comment]
> Is it better if we just take advantage of the standard `ClientCapabilities`?
i.e. 
`config.GocodeCompletionEnabled` -> `init.Capabilities.TextDocument.Completion`
`config.FuncSnippetEnabled` -> `init.Capabilities.TextDocument.Completion.SnippetSupport`

cc @ramya-rao-a

**Comment by ramya-rao-a** [comment]
_TLDR_; 

That works for me. It will cause double completions to show up for users who have updated the language server but not the Go extension to VS Code which they can solve by either disabling the language server or updating the Go extension.

_Long version:_

The Go extension for VS Code has no control over the client capabilities being sent to the language server.

`init.Capabilities.TextDocument.Completion` will always be `true` for any VS Code extension that uses the [VS Code Language Client](https://github.com/Microsoft/vscode-languageserver-node/tree/master/client) today. The Client Capabilities only differ from version to version of VS Code client. For example, the `Go to Implementation` feature was a new capability added recently and will not be available in the older versions of the VS Code client.

I'll have to edit the Go extension (and release an update) to not register its own completion provider when the language provider provides one. Users who have updated the language server but not the Go extension, will get double completions.

**Comment by keegancsmith** [comment]
> The Go extension for VS Code has no control over the client capabilities being sent to the language server.

Then lets stick to using the initializationOptions as in this PR. I'll merge this in.

**Pull request merged**